pkg/sql/drop_table.go

```diff
@@ -241,22 +241,97 @@ func (p *planner) canRemoveInterleave(
 	return p.CheckPrivilege(ctx, table, privilege.CREATE)
 }
 
-func (p *planner) removeInterleave(ctx context.Context, ref descpb.ForeignKeyReference) error {
-	table, err := p.Descriptors().GetMutableTableVersionByID(ctx, ref.Table, p.txn)
-	if err != nil {
-		return err
-	}
-	if table.Dropped() {
+// removeAncestorReferences removes any references to `idx` from `descendentIdx`
+// list of ancestors. `descendentIdx` should be an interleaved descendent of
+// `idx`. `idx` should belong to `tableDesc` and `descendentIdx` should belong
+// to `descendentTable`.
+func (p *planner) removeAncestorReferences(
+	ctx context.Context,
+	tableDesc *tabledesc.Mutable,
+	idx *descpb.IndexDescriptor,
+	descendentTable *tabledesc.Mutable,
+	descendentIdx *descpb.IndexDescriptor,
+) error {
+
+	if descendentTable.Dropped() {
 		// The referenced table is being dropped. No need to modify it further.
 		return nil
 	}
-	idx, err := table.FindIndexByID(ref.Index)
-	if err != nil {
-		return err
+
+	foundAncestor := false
+	for k, ancestor := range descendentIdx.Interleave.Ancestors {
+		if ancestor.TableID == tableDesc.ID && ancestor.IndexID == idx.ID {
+			if foundAncestor {
+				return errors.AssertionFailedf(
+					"ancestor entry in %s for %s@%s found more than once",
+					tableDesc.Name, tableDesc.Name, idx.Name)
+			}
+			descendentIdx.Interleave.Ancestors = append(descendentIdx.Interleave.Ancestors[:k],
+				descendentIdx.Interleave.Ancestors[k+1:]...)
+			foundAncestor = true
+		}
 	}
-	idx.Interleave.Ancestors = nil
+
 	// No job description, since this is presumably part of some larger schema change.
-	return p.writeSchemaChange(ctx, table, descpb.InvalidMutationID, "")
+	return p.writeSchemaChange(ctx, descendentTable, descpb.InvalidMutationID, "")
+}
+
+// removeInterleaveReferences removes interleaved relationships from both the
+// given index's parents and descendents.
+func (p *planner) removeInterleaveReferences(
```

Comment:
This actually isn't quite right. We also want to remove all of the ancestor's of the index that we want to drop from all of it's descendants. Will update.

Reply:
Updated this to remove the prefix of ancestors, which looks like it's doing the right thing. Updating with some tests in SQL land the exercise this.